### PR TITLE
add detection confusion matrix

### PR DESCRIPTION
**Pull request by nisyad-ms**
This PR adds a new metric DetectionConfusionMatrix which calculates the confusion matrix in an object detection setting.

### CODE REVIEW

File: visionmetrics/detection/confusion_matrix.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        self.add_state("fp_due_to_no_corresponding_gt_box", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("confusion_matrix", default=[], dist_reduce_fx=None)
+
+    def update(self, predictions, targets):
+        if len(predictions) != len(targets):
+            raise ValueError("Number of predictions and targets should be the same.")
+        if not isinstance(predictions[0][0], list):
+            raise ValueError(f"Expected predictions to be a list of lists, got a list of {type(predictions[0][0])}")
+
+        self.confusion_matrix.append(self._update_confusion_matrix(predictions, targets, self.iou_threshold))
```
Comment (shonohs) [inline visionmetrics/detection/confusion_matrix.py:45]: maybe we don't need to construct the dictionary here, just update the tp*, fp*. then create a dictionary in `compute()` method?

Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: Great suggestion! Updated.

**Review by shonohs — APPROVED** [review]
